includes/class-indieauth-token-endpoint.php

```diff
@@ -0,0 +1,225 @@
+			'indieauth/1.0', '/token', array(
+				array(
+					'methods'  => WP_REST_Server::CREATABLE,
+					'callback' => array( $this, 'post' ),
+					'args'     => array(
+						'grant_type'   => array(),
+						'code'         => array(),
+						'client_id'    => array(
+							'validate_callback' => array( $this, 'is_valid_url' ),
+							'sanitize_callback' => 'esc_url_raw',
+						),
+						'redirect_uri' => array(
+							'validate_callback' => array( $this, 'is_valid_url' ),
+							'sanitize_callback' => 'esc_url_raw',
+						),
+						'me'           => array(
+							'validate_callback' => array( $this, 'is_valid_url' ),
+							'sanitize_callback' => 'esc_url_raw',
+						),
+						'action'       => array(),
+						'token'        => array(),
+					),
+				),
+			)
+		);
+		register_rest_route(
+			'indieauth/1.0', '/token', array(
+				array(
+					'methods'  => WP_REST_Server::READABLE,
+					'callback' => array( $this, 'get' ),
+					'args'     => array(),
+				),
+			)
+		);
+	}
+
+	/**
+	 * Returns if valid URL for REST validation
+	 *
+	 * @param string $url
+	 *
+	 * @return boolean
+	 */
+	public static function is_valid_url( $url, $request = null, $key = null ) {
+		if ( ! is_string( $url ) || empty( $url ) ) {
+			return false;
+		}
+		return filter_var( $url, FILTER_VALIDATE_URL );
+	}
+
+	public function generate() {
+		return wp_generate_password( 128, false );
+	}
+
+	public function get_token( $token, $hash = true ) {
+		$key = '_indieauth_token_';
+		// Either token is already hashed or is not
+			$key .= $hash ? md5( $token ) : $token;
+		$args     = array(
+			'number'      => 1,
+			'count_total' => false,
+			'meta_query'  => array(
+				array(
+					'key'     => $key,
+					'compare' => 'EXISTS',
+				),
+			),
+		);
+		$query    = new WP_User_Query( $args );
+		$results  = $query->get_results();
+		if ( empty( $results ) ) {
+			return null;
+		}
+		$user  = $results[0];
+		$value = get_user_meta( $user->ID, $key, true );
+		if ( empty( $value ) ) {
+				return null;
+		}
+		$value['user'] = $user->ID;
+		return $value;
+	}
+
+	public function get( $request ) {
+		$params       = $request->get_params();
+		$access_token = self::get_token_from_bearer_header( $request->get_header( 'Authorization' ) );
+		if ( ! $access_token ) {
+			return new WP_Error(
+				'invalid_request', __( 'Bearer Token Not Supplied', 'indieauth' ),
+				array(
+					'status' => '401',
+				)
+			);
+		}
+		$token = self::get_token( $access_token );
+		if ( ! $token ) {
+			return new WP_Error(
+				'invalid_access_token', __( 'Invalid Token', 'indieauth' ),
+				array(
+					'status'   => '401',
+					'response' => $access_token,
+				)
+			);
+		}
+		return rest_ensure_response( $token );
+	}
+
+	public function set_token( $token ) {
+		// Token consists of properties at minimum: access_token, me, scope
+		if ( ! isset( $token['access_token'] ) || ! isset( $token['me'] ) ) {
+			return false;
+		}
+		$access_token = md5( $token['access_token'] );
```

Comment:
Should probably use a Wordpress hashing function rather than plain md5, maybe this one? https://codex.wordpress.org/Function_Reference/wp_hash_password